### PR TITLE
feat: HTTP API

### PR DESCRIPTION
**Pull request by alvancamp**
This PR adds an (undocumented) MVP HTTP REST API to SuperConductor. The plan for the future is to use something like Swagger combined with JSDoc to generate full API documentation, but this gets us going for now.

The API is dynamically generated at runtime by iterating over the IPCServer methods, abiding by the following rules:

- All IPCServer methods must accept only a single argument and that argument must be an object. This is because `request.query` is passed directly to the IPCServer method as its only argument. The codebase has been changed to accommodate this rule. I don't think this is a security risk but I'd like a second opinion.
- `GET` methods must be prefixed with `get`.
- `DELETE` methods must be prefixed with `delete`.
- All other methods are assumed to be `POST`.
- Methods beginning with `_` are not exposed.

There's a few gotchas with this implementation:
- It doesn't always return the expected status code. If some arguments are missing, it'll most likely return a 404 or a 500. It isn't smart enough to know when to return a 400. This could potentially be fixed with the addition of full JSDoc comments and codegen, I'm not 100% sure though.
- The port is hardcoded to 5500. We could of course add something that allows this to be changed, but I'm not quite sure where we'd put it in the UI.

Example uses:

- Get a Rundown: `GET` `http://localhost:5500/api/rundown?rundownId=default.rundown.json`
- Play a Part: `POST` to `http://localhost:5500/api/playPart?rundownId=default.rundown.json&groupId=uVxbR8Gc&partId=1tuy9Y1W`
- Log an error message of `foo bar` (note the multiple usage of `params`): `POST` to `http://localhost:5500/api/log?level=error&params=foo&params=bar`

### CODE REVIEW

**Review by nytamin — COMMENTED** [review]
Looks pretty good!

I pushed some minor changes, adding an index page at `/api`.

Todo (?)

- [ ] Update README
- [ ] Is there a way to extract the signatures of the methods, so that the parameter names and types can be exposed somehow?

**Comment by nytamin** [comment]
>The port is hardcoded to 5500

We should allow changing it through CLI, here: https://github.com/SuperFlyTV/SuperConductor/blob/feat/http-api/apps/app/src/electron/SuperConductor.ts#L105

File: apps/app/src/electron/HTTPAPI.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+				endpoint = methodName.charAt(3).toLocaleLowerCase() + methodName.substring(4)
+				endpointType = 'GET'
+				this.router.get(`/api/${endpoint}`, async (ctx) => {
+					try {
+						const result = await fcn(ctx.request.query)
+						ctx.response.body = result
+						ctx.response.status = 200
+					} catch (error) {
+						const stringifiedError = stringifyError(error)
+						if (stringifiedError.includes('not found')) {
```
Comment (nytamin) [inline apps/app/src/electron/HTTPAPI.ts:65]: It is slightly better to to `.match(/not found/i)` to be case insensitive

**Comment by nytamin** [comment]
Also, something that I think we should do is to treat this as an "internal" API for now. By exposing all these methods we're going to have a very hard time in the future when we want to change something internally.

So I think we should expose this under `/api/internal/xxxx` for now, and then we should start thinking about making a Good API which is explicitly defined, documented and stable, some time later.

File: apps/app/src/electron/HTTPAPI.ts (hunk 1 above)
Comment (alvancamp) [reply apps/app/src/electron/HTTPAPI.ts]: Resolved in https://github.com/SuperFlyTV/SuperConductor/pull/83/commits/971741312f0e4235319e0ebf471d6211cb839d54

**Comment by nytamin** [comment]
> I believe I have addressed all the above feedback.
> 
> > * [ ]   Is there a way to extract the signatures of the methods, so that the parameter names and types can be exposed somehow?
> 
> As far as I know, the only way to achieve this would be to use the TypeScript parser to walk the AST and then do code generation, since type information isn't available at runtime. Argument names can technically be extracted by stringifying the function and then parsing that, but it's messy and probably not something we should rely on. I've done some code generation in the past, but not specifically from an AST like this.

Ok, we'll look into that for The Good And Stable API later then

**Pull request merged**